app/waf/src/controllers/adc.controller.ts

```diff
@@ -732,74 +689,155 @@ export class AdcController extends BaseController {
-      check: Promise.resolve(this.reclaimed()),
+      check: this.reclaimed,
       next: [AdcState.POWERON],
     },
+    {
+      state: AdcState.ACTIVE,
+      check: (ctrl: AdcStateCtrlr) => {
+        return Promise.resolve(true);
+      },
+      next: [AdcState.RECLAIMED],
+    },
   ];
 
   constructor(private adc: Adc, private addon: AddonReqValues) {}
 
   async readyTo(state: string): Promise<boolean> {
     let stateEntry = this.getStateEntry(this.adc.status);
-    return (await stateEntry.check) && stateEntry.next.includes(state);
+
+    if (this.adc.status.endsWith('ERROR')) {
+      if (state === stateEntry.state) return true;
+      else return false;
+    }
+
+    return stateEntry.next.includes(state) && (await stateEntry['check'](this));
   }
 
   async gotTo(state: string): Promise<boolean> {
-    return this.getStateEntry(state).check;
+    return this.getStateEntry(state)['check'](this);
   }
 
   private getStateEntry(name: string): AdcStateEntry {
     return this.states.find(s => {
-      return s.state === name;
+      return s.state === name || s.failure === name;
     })!;
   }
 
-  private reclaimed(): boolean {
-    if (this.adc.management) return false;
-    if (this.adc.compute.vmId) return false;
-    for (let net of Object.keys(this.adc.networks)) {
-      if (this.adc.networks[net].portId) return false;
-    }
-    return true;
+  private async getBigipMgr(): Promise<BigIpManager> {
+    if (!this.adc.management)
+      throw new Error(
+        `The management session of ADC is empty, cannot initialize bigip manager.`,
+      );
+
+    let mgmt = this.adc.management;
+    return BigIpManager.instanlize({
+      username: mgmt.username,
+      password: mgmt.password,
+      ipAddr: mgmt.ipAddress,
+      port: mgmt.tcpPort,
+    });
   }
 
-  private async trusted(): Promise<boolean> {
-    return Promise.resolve(true);
+  private async getAsgMgr(): Promise<ASGManager> {
+    let svc = await new ASGServiceProvider().value();
+    return new ASGManager(svc);
   }
 
-  private async onboarded(): Promise<boolean> {
-    let bigipMgr = await this.getBigipMgr();
+  // Notices:
+  // Why not use 'this' in the following functions:
+  // The scope of 'this' changes when placing 'this.onboarded/trusted/...'
+  // in the AdcStateEntry declaration in the above 'states' member.
+  // So, please pass in the 'ctrl' object to when accessing AdcStateCtrlr.
+
+  private reclaimed(ctrl: AdcStateCtrlr): Promise<boolean> {
+    return Promise.resolve(
+      ((): boolean => {
+        if (ctrl.adc.management) return false;
+        if (ctrl.adc.compute.vmId) return false;
+        for (let net of Object.keys(ctrl.adc.networks)) {
+          if (ctrl.adc.networks[net].portId) return false;
+        }
+        return true;
+      })(),
+    );
+  }
+
+  private async trusted(ctrl: AdcStateCtrlr): Promise<boolean> {
+    if (!ctrl.adc.trustedDeviceId!) return false;
+
+    let asgMgr = await ctrl.getAsgMgr();
```

Comment:
let asgMgr = ctrl.asgMgr;

Reply:
It's not a member of ctrl.

Reply:
Sorry, I used to assume the ctrl is adc controller.